src/SDKs/Batch/DataPlane/Azure.Batch/AddTasksWorkflowManager.cs

```diff
@@ -293,18 +295,55 @@ private bool IsWorkflowDone()
             //
             // Fire the protocol add collection request
             //
-            var asyncTask = this._jobOperations.ParentBatchClient.ProtocolLayer.AddTaskCollection(
-                this._jobId,
-                protoTasksToAdd,
-                this._behaviorManager,
-                this._parallelOptions.CancellationToken);
-
-            var response = await asyncTask.ConfigureAwait(continueOnCapturedContext: false);
-
-            //
-            // Process the results of the add task collection request
-            //
-            this.ProcessProtocolAddTaskResults(response.Body.Value, tasksToAdd);
+            try
+            {
+                var asyncTask = this._jobOperations.ParentBatchClient.ProtocolLayer.AddTaskCollection(
+                    this._jobId,
+                    protoTasksToAdd,
+                    this._behaviorManager,
+                    this._parallelOptions.CancellationToken);
+
+                var response = await asyncTask.ConfigureAwait(continueOnCapturedContext: false);
+                //
+                // Process the results of the add task collection request
+                //
+                this.ProcessProtocolAddTaskResults(response.Body.Value, tasksToAdd);
+            }
+            catch(Common.BatchException e)
+            {
+                if (e.InnerException is Models.BatchErrorException)
+                {
+                    Models.BatchError error = ((Models.BatchErrorException) e.InnerException).Body;
+                    if (error.Code == Common.BatchErrorCodeStrings.RequestBodyTooLarge)
+                    {
+                        //
+                        // Our chunk sizes were too large, so universally reduce size
+                        // This is an internal error due to us using greedy initial maximum chunk size,
+                        //   so do not increment retry counter.
+                        //
+                        int currLength = tasksToAdd.Count;
+                        // If we are already submitting only a single task we cannot reduce in size any further
+                        if (currLength == 1)
+                        {
+                            throw e;
+                        }
+                        int newLength = currLength / 2;
+                        Interlocked.CompareExchange(ref this._maxTasks, newLength, currLength);
+                        foreach (TrackedCloudTask trackedTask in tasksToAdd.Values)
+                        {
+                            this._remainingTasksToAdd.Enqueue(trackedTask);
+                        }
+                    }
+                    else
+                    {
+                        throw e;
```

Comment:
If this stays here, it should just be `throw` -- and I'd just drop all the else's and let it fall through to 1 `throw` at the bottom.